samples/security/TokenLoginModule.java

```diff
@@ -0,0 +1,250 @@
+/**
+ * Copyright (c) 2007-2017, Kaazing Corporation. All rights reserved.
+ */
+
+package com.kaazing.demo.loginmodules;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import javax.security.auth.Subject;
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.auth.login.LoginException;
+import javax.security.auth.spi.LoginModule;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.kaazing.gateway.server.spi.security.AuthenticationToken;
+import org.kaazing.gateway.server.spi.security.AuthenticationTokenCallback;
+import org.kaazing.gateway.server.spi.security.ExpiringState;
+import org.kaazing.gateway.server.spi.security.LoginModuleOptions;
+import org.kaazing.gateway.server.spi.security.LoginResult;
+import org.kaazing.gateway.server.spi.security.LoginResultCallback;
+
+// This is an example custom login module that reads a token to extract the username. If
+// successful, the username is placed into shared state for login modules downstream in
+// the chain to continue authorization. For example, a downstream login module may retrieve
+// the user's role or roles from LDAP and inject them into the Subject.
+//
+// A good starting point for understanding login modules and recommended practices is
+// the Java Authentication and Authorization Service (JAAS): LoginModule Developer's Guide:
+//
+// http://docs.oracle.com/javase/8/docs/technotes/guides/security/jaas/JAASLMDevGuide.html
+
+public class TokenLoginModule implements LoginModule {
+
+    private static final String USERNAME_KEY = "javax.security.auth.login.name";
+
+    private Subject subject;
+    private CallbackHandler handler;
+    private Map<String, Object> sharedState;
+
+    private ExpiringState expiringState;
+
+    private String username;
+
+    // Flag to enable or disable debugging output.
+    private boolean debug = false;
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState, Map<String, ?> options) {
+
+        this.debug = ((String) options.get("debug")).equalsIgnoreCase("true");
```

Comment:
This should be `"true".equalsIgnoreCase((String) options.get("debug"))` to avoid a `NullPointerException` is the option is not set.